docs/design/speed-test.md

```diff
@@ -0,0 +1,210 @@
+# Upload Speed Test for Backup Storage Location (Object Storage)
+
+## Abstract
+
+This document presents the design of a Custom Resource Definition (CRD) and its controller to test the upload speed from an OpenShift cluster to cloud object storage.
+The design leverages the BackupStorageLocation configuration from the OADP operator’s DPA CRD to specify object storage locations for testing. The controller will use
+these configurations to authenticate and perform uploads, measuring speed and reporting results through the CRD status.
+
+## Background
+
+We have observed several instances in customer environments where backup or restore operations became stalled or significantly delayed, but the root cause was not immediately apparent.
+These issues could stem from poor network connectivity, cloud storage throttling, or misconfigurations in BackupStorageLocation (BSL) settings. Since backup performance plays a critical
+role in disaster recovery strategies, it is essential to have a reliable way to test upload speeds.
+
+This Upload Speed Test CRD will help in such scenarios by proactively measuring the upload speed from the OpenShift cluster to the cloud storage configured in OADP’s BSL. By identifying
+performance bottlenecks ahead of backup or restore operations, administrators can take corrective actions to avoid failures during critical operations.
+
+## Goals
+
+-   Measure upload speeds for a given BackupStorageLocation (BSL) config settings and credentials
+-   Measure upload speeds from the OpenShift cluster to the configured object storage.
+-   Update the test results in the status of the UploadSpeedTest CRD for visibility and tracking.
+-   Enable smooth integration with OADP operator-managed backup storage location configurations.
+
+
+## Non-Goals
+
+- This design will not create or modify BackupStorageLocation entries in OADP.
+- It will not implement download or latency tests, focusing solely on upload speed.
+- Scheduling of recurring tests is not supported in the initial version.
+
+## High-Level Design
+Components involved and their responsibilities:
+
+- UploadSpeedTest (UST) CRD:
```

Comment:
Reading whole design, would make sense to change CRD name to something more generic (NetworkSpeedTest?). If more features, like latency and download speed, are added, the CRD name would not be telling users all things it can do?

Reply:
Yeah I had thought about that, Any Network speed would consist of upload as well as download speed, the risk here with `NetworkSpeedTest` is that it might give a false impression that we are catering/calculating download speed as well, but thats not our goal here. Hence `UploadSpeedTest`.